Xamarin.PropertyEditing.Mac/Controls/BasePathEditorControl.cs

```diff
@@ -0,0 +1,148 @@
+{
+	internal abstract class BasePathEditorControl<T> : PropertyEditorControl<PropertyViewModel<T>>
+	{
+		private const string PathRevealName = "pe-path-reveal";
+		private const string PathBrowseName = "pe-path-browse";
+
+		protected readonly TextFieldSmallButtonContainer currentTextField;
+
+		protected readonly NSOpenPanel panel;
+		protected readonly SmallButton browsePathButton;
+		protected readonly SmallButton revealPathButton;
+
+		public override NSView FirstKeyView => this.currentTextField;
+		public override NSView LastKeyView => this.revealPathButton.Enabled ? this.revealPathButton : this.browsePathButton;
+
+		private readonly NSObject[] objects;
+		public override NSObject[] AccessibilityChildren { get => this.objects; }
+
+		protected BasePathEditorControl (IHostResourceProvider hostResource)
+			: base (hostResource)
+		{
+			this.currentTextField = new TextFieldSmallButtonContainer ();
+			this.currentTextField.Changed += CurrentTextField_Changed;
+			AddSubview (this.currentTextField);
+
+			#region Reveal handler
+
+			this.revealPathButton = new SmallButton {
+				Font = NSFont.FromFontName (DefaultFontName, DefaultFontSize),
+				Image = HostResources.GetNamedImage (PathRevealName),
+				StringValue = string.Empty
+			};
+
+			this.currentTextField.AddButton (this.revealPathButton);
+
+			this.panel = new NSOpenPanel {
+				ShowsResizeIndicator = true,
+				ShowsHiddenFiles = false,
+				CanCreateDirectories = true,
+				AllowsMultipleSelection = false
+			};
+
+			// update the value on keypress
+			this.revealPathButton.Activated += OnRevealPathButtonActivated;
+
+			#endregion
+
+			#region Browse Path
+
+			this.browsePathButton = new SmallButton {
+				Font = NSFont.FromFontName (DefaultFontName, DefaultFontSize),
+				Image = HostResources.GetNamedImage (PathBrowseName),
+				StringValue = string.Empty
+			};
+			this.browsePathButton.Activated += BrowsePathButton_Activated;
+			this.currentTextField.AddButton (this.browsePathButton);
+
+			this.objects = new NSObject[3];
+			this.objects[0] = this.currentTextField;
+			this.objects[1] = this.browsePathButton;
+			this.objects[2] = this.revealPathButton;
+			this.AccessibilityElement = true;
+			this.AccessibilityRole = NSAccessibilityRoles.GroupRole;
+
+			this.currentTextField.AccessibilityRole = NSAccessibilityRoles.TextFieldRole;
+			this.revealPathButton.AccessibilityRole = NSAccessibilityRoles.ButtonRole;
+			this.browsePathButton.AccessibilityRole = NSAccessibilityRoles.ButtonRole;
+
+			#endregion
+
+			AddConstraints (new[] {
+				NSLayoutConstraint.Create (this.currentTextField, NSLayoutAttribute.Left, NSLayoutRelation.Equal, this, NSLayoutAttribute.Left, 1, 0f),
+				NSLayoutConstraint.Create (this.currentTextField, NSLayoutAttribute.CenterY, NSLayoutRelation.Equal, this, NSLayoutAttribute.CenterY, 1, 0f),
+				NSLayoutConstraint.Create (this.currentTextField, NSLayoutAttribute.Width, NSLayoutRelation.Equal, this, NSLayoutAttribute.Width, 1, -33f),

```

Comment:
```suggestion
				NSLayoutConstraint.Create (this.currentTextField, NSLayoutAttribute.Width, NSLayoutRelation.Equal, this, NSLayoutAttribute.Width, 1, 0f),
```
As of #546  `EditorContainer` now handles the property button, so we don't need to size for it.